### PR TITLE
udp downstream api - work in progress

### PR DESCRIPTION
**Pull request by alonbg**
[nginx.conf](https://github.com/openresty/stream-lua-nginx-module/files/386537/nginx.txt) with both tcp and udp test servers

testing udp: not working ( just yet ... )
client: `socat -T10 -t10 -d -d -d - udp:127.0.0.1:8070`
echo server: `socat -d -d -d  udp-l:8071,reuseaddr,fork,crlf system:"cat"`

testing tcp: working just fine
client: `socat -d -d -d - tcp:127.0.0.1:5354`
echo server: `socat -d -d -d  tcp-l:5355,reuseaddr,fork,crlf system:"cat"

### CODE REVIEW

File: src/ngx_stream_lua_socket_tcp.c [hunk 1]
```diff
@@ -3919,36 +3911,44 @@ ngx_stream_lua_socket_cleanup_compiled_pattern(lua_State *L)
...
+
 static int
-ngx_stream_lua_req_socket(lua_State *L)
+ngx_stream_lua_tcp_req_socket(lua_State *L)
 {
-    int                              n, raw;
-    ngx_peer_connection_t           *pc;
+	int                              n, raw;
+	ngx_stream_session_t			*s;
+	ngx_peer_connection_t           *pc;
```
Comment (agentzh) [inline src/ngx_stream_lua_socket_tcp.c]: Style: indentation problems here?

Comment (agentzh) [inline src/ngx_stream_lua_socket_tcp.c]: Please fix the indentation issues in the rest of your changes. We use 4-space indentation consistently.

File: src/ngx_stream_lua_socket_tcp.c [hunk 2]
```diff
@@ -3957,6 +3957,11 @@ ngx_stream_lua_req_socket(lua_State *L)
 
     c = s->connection;
 
+    if ( SOCK_STREAM != c->type) {
```
Comment (agentzh) [inline src/ngx_stream_lua_socket_tcp.c]: Coding style problems here. It should be

```C
if (c->type != SOCK_STREAM) {
```

instead.

File: src/ngx_stream_lua_socket_tcp.c [hunk 3]
```diff
@@ -3957,6 +3957,11 @@ ngx_stream_lua_req_socket(lua_State *L)
...
+    	return luaL_error(L, "socket api does not match connection transport",
+    					  lua_gettop(L));
```
Comment (agentzh) [inline src/ngx_stream_lua_socket_tcp.c]: The continued line should be aligned with the first function call argument in the previous line. See the rest of the code base for examples. BTW, please use the `ngx-releng` tool from `nginx-devel-utils` to check coding style issues.

**Comment by agentzh** [comment]
Will you add some test cases to the existing test suite? Also, please rebase to the current git master branch in your own branch. Many thanks!

File: src/ngx_stream_lua_socket_udp.c [hunk 4]
```diff
@@ -83,7 +90,7 @@ ngx_stream_lua_inject_socket_udp_api(ngx_log_t *log, lua_State *L)
     lua_setfield(L, -2, "udp"); /* ngx socket */
 
     /* udp socket object metatable */
-    lua_pushlightuserdata(L, &ngx_stream_lua_socket_udp_metatable_key);
+    lua_pushlightuserdata(L, &ngx_stream_lua_udp_socket_metatable_key);
```
Comment (agentzh) [inline src/ngx_stream_lua_socket_udp.c]: This name change is confusing. We regard `ngx_stream_lua_socket_` as something like a namespace.

File: src/ngx_stream_lua_socket_udp.h [hunk 5]
```diff
@@ -46,18 +46,19 @@ struct ngx_stream_lua_socket_udp_upstream_s {
...
-    size_t                           recv_buf_size;
+    ngx_uint_t                         ft_type;
+    ngx_err_t                          socket_errno;
+    size_t                             received; /* for receive */
+    size_t                             recv_buf_size;
 
     ngx_stream_lua_co_ctx_t           *co_ctx;
-
-    unsigned                         waiting; /* :1 */
+    unsigned                           raw_downstream:1;
```
Comment (agentzh) [inline src/ngx_stream_lua_socket_udp.h]: We don't really need this field since the downstream cosockets in the stream subsystem should always be "raw" anyway.

File: src/ngx_stream_lua_socket_udp.c (hunk 4 above)
Comment (alonbg) [inline src/ngx_stream_lua_socket_udp.c]: it matchs [tcp](https://github.com/openresty/stream-lua-nginx-module/blob/master/src/ngx_stream_lua_socket_tcp.c#L196).
I can change both :)

File: src/ngx_stream_lua_socket_tcp.c (hunk 3 above)
Comment (alonbg) [inline src/ngx_stream_lua_socket_tcp.c]: done.
Would be very helpful though adding to nginx-devel:
1. [Install ack-grep](http://beyondgrep.com/install)
2. Add nginx-devel to your run PATH
:)

File: src/ngx_stream_lua_socket_udp.h (hunk 5 above)
Comment (alonbg) [inline src/ngx_stream_lua_socket_udp.h]: it matchs [tcp](https://github.com/openresty/stream-lua-nginx-module/blob/master/src/ngx_stream_lua_socket_tcp.h#L104)
I can change tcp as well :)

File: src/ngx_stream_lua_socket_udp.c (hunk 4 above)
Comment (alonbg) [inline src/ngx_stream_lua_socket_udp.c]: Not sure  Test::Nginx has proper support for udp.
[new issue under test-nginx](https://github.com/openresty/test-nginx/issues/40)

File: src/ngx_stream_lua_socket_tcp.c (hunk 3 above)
Comment (agentzh) [inline src/ngx_stream_lua_socket_tcp.c]: @alonbg I agree. Patches welcome to nginx-devel-utils! :)

File: src/ngx_stream_lua_socket_tcp.c (hunk 2 above)
Comment (agentzh) [inline src/ngx_stream_lua_socket_tcp.c]: Please note the extra space after `(` in this line, which should get removed :)

File: src/ngx_stream_lua_socket_udp.c (hunk 4 above)
Comment (agentzh) [inline src/ngx_stream_lua_socket_udp.c]: @alonbg Yeah, we need to add UDP support to that test class :) Patches welcome! :)

File: src/ngx_stream_lua_socket_udp.h (hunk 5 above)
Comment (agentzh) [inline src/ngx_stream_lua_socket_udp.h]: @alonbg Ah, I was not aware of that. Never mind then :)

**Comment by alonbg** [comment]
Rebased.
Travis fails on udp tests requiring Nginx::Socket::Lua::Dgram ...

File: src/ngx_stream_lua_socket_udp.c [hunk 6]
```diff
@@ -106,18 +113,206 @@ ngx_stream_lua_inject_socket_udp_api(ngx_log_t *log, lua_State *L)
...
+    }
+
+    ngx_stream_lua_check_context(L, ctx, NGX_STREAM_LUA_CONTEXT_CONTENT);
+
+    c = s->connection;
+
+    if (c->type != SOCK_DGRAM) {
+        return luaL_error(L, "socket api does not match connection transport",
+                          lua_gettop(L));
+        }
```
Comment (agentzh) [inline src/ngx_stream_lua_socket_udp.c]: Indentation problem here?

File: src/ngx_stream_lua_util.c [hunk 7]
```diff
@@ -3227,9 +3227,8 @@ ngx_stream_lua_inject_req_api(ngx_log_t *log, lua_State *L)
     /* ngx.req table */
 
     lua_createtable(L, 0 /* narr */, 1 /* nrec */);    /* .req */
```
Comment (agentzh) [inline src/ngx_stream_lua_util.c]: Need to change the number `1` here to `2` as well. Otherwise the test in `062-count.t` would be meaningless since the numbers are out of sync.

File: src/ngx_stream_lua_util.c [hunk 8]
```diff
@@ -3227,9 +3227,8 @@ ngx_stream_lua_inject_req_api(ngx_log_t *log, lua_State *L)
...
-
-    ngx_stream_lua_inject_req_socket_api(L);
-
+    ngx_stream_lua_inject_tcp_req_socket_api(L);
+    ngx_stream_lua_inject_udp_req_socket_api(L);
     lua_setfield(L, -2, "req");
```
Comment (agentzh) [inline src/ngx_stream_lua_util.c]: Need a blank line before this line.

File: src/ngx_stream_lua_socket_udp.c [hunk 9]
```diff
@@ -1107,6 +1302,26 @@ ngx_stream_lua_socket_udp_upstream_destroy(lua_State *L)
...
+        return 0;
+    }
+
+    if (u->cleanup) {
+        ngx_stream_lua_socket_udp_cleanup(u); /* it will clear u->cleanup */
+    }
+
+    return 0;
+}
+
```
Comment (agentzh) [inline src/ngx_stream_lua_socket_udp.c]: Style: need 2 blank lines here.

**Comment by alonbg** [comment]
all comments committed. rebased.
Test fail on `Can't locate Test/Nginx/Socket/Lua/Dgram.pm`
Please see test_nginx [pull request](https://github.com/openresty/test-nginx/pull/41) for that

**Comment by alonbg** [comment]
@agentzh tests still fail. Please advise :)

```
t/137-udp-count.t
#   Failed test 'TEST 2: entries under the metatable of req sockets - status code ok'
#   at /home/alon/perl5/lib/perl5/Test/Nginx/Socket.pm line 895.
#          got: ''
#     expected: '200'

#   Failed test 'TEST 2: entries under the metatable of req sockets - response_body - response is expected (repeated req 0, req 0)'
#   at /home/alon/perl5/lib/perl5/Test/Nginx/Socket.pm line 1341.
#          got: ''
#     expected: 'n = 4
# '


138-req-udp-socket.t:
TEST 1: sanity - Can't connect to 127.0.0.1:1984: Connection refused
	Retry connecting after 0.675 sec
TEST 1: sanity - Can't connect to 127.0.0.1:1984: Connection refused
	Retry connecting after 0.825 sec
TEST 1: sanity - Can't connect to 127.0.0.1:1984: Connection refused
	Retry connecting after 0.99 sec

.
.
```

**Comment by agentzh** [comment]
@alonbg The "Connection refused" error reported by the test scaffold is a clear indicator of nginx server crashes (segmentation faults). In this case, we should run that offending test block individually with the valgrind test mode to pinpoint the memory issue in the code base.

**Comment by agentzh** [comment]
@alonbg I'm also getting valgrind memory errors while running your tests on my side. For example,

```
TEST 1: sanity
==106850== Jump to the invalid address stated on the next line
==106850==    at 0x0: ???
==106850==    by 0x41FAF8: ngx_chain_writer (ngx_output_chain.c:741)
==106850==    by 0x4FD59E: ngx_stream_lua_send_chain_link (ngx_stream_lua_output.c:716)
==106850==    by 0x4FDA53: ngx_stream_lua_ngx_echo (ngx_stream_lua_output.c:227)
==106850==    by 0x4FDB25: ngx_stream_lua_ngx_say (ngx_stream_lua_output.c:36)
==106850==    by 0x5465177: lj_BC_FUNCC (in /opt/luajit21sysm/lib/libluajit-5.1.so.2.1.0)
==106850==    by 0x4EF88C: ngx_stream_lua_run_thread (ngx_stream_lua_util.c:1220)
==106850==    by 0x4EDD47: ngx_stream_lua_content_by_chunk (ngx_stream_lua_contentby.c:167)
==106850==    by 0x4EDEC1: ngx_stream_lua_content_handler_inline (ngx_stream_lua_contentby.c:74)
==106850==    by 0x4EE0CE: ngx_stream_lua_content_handler (ngx_stream_lua_contentby.c:236)
==106850==    by 0x49776A: ngx_stream_init_session (ngx_stream_handler.c:244)
==106850==    by 0x497D94: ngx_stream_init_connection (ngx_stream_handler.c:223)
==106850==    by 0x43B1B1: ngx_event_recvmsg (ngx_event_accept.c:625)
==106850==    by 0x443502: ngx_epoll_process_events (ngx_epoll_module.c:900)
==106850==    by 0x439C1F: ngx_process_events_and_timers (ngx_event.c:242)
==106850==    by 0x442558: ngx_single_process_cycle (ngx_process_cycle.c:309)
==106850==    by 0x41BA20: main (nginx.c:364)
==106850==  Address 0x0 is not stack'd, malloc'd or (recently) free'd
==106850==
```

**Comment by agentzh** [comment]
@alonbg BTW, please rebase to the latest git master for some travis ci fixes (due to recent travis ci server environment changes. alas).

**Comment by alonbg** [comment]
@agentzh
`ngx_chain_xxx` are used from `ngx_stream_lua_ngx_echo (ngx_stream_lua_output.c:227)` downwards. This goes for `ngx.say()`, `ngx.print()` and `ngx.flush()`.
Can we still use the chain memory management api for udp ?
How do you suggest we handle this ?
Thanks!

**Comment by agentzh** [comment]
@alonbg I think you encountered this while working on ngx_stream_echo_module as well. You cannot use `ngx_chain_writer` for UDP sends at all. You ended up with calling `ngx_udp_send` in ngx_stream_echo, remember?

**Comment by alonbg** [comment]
@agentzh yes I remembered that :) 
Just hoped for an easier way than refactoring this area. 
BTW, I do need to allocate there.  What api should I use? 
Thanks

**Comment by agentzh** [comment]
@alonbg What API should you use? Sorry, I don't know what kind of refactoring are you using. I only have the time to comment on your code changes atm. Please see the existing code for a better idea of how to use the nginx C API (and also those nginx programming tutorials out there on the web).

**Comment by agentzh** [comment]
@alonbg I suggest you skip those `ngx.print`, `ngx.say` API functions for the UDP downstream since they make very little sense for this context.

**Comment by alonbg** [comment]
Thanks @agentzh .  Going through those tutorials as we speak :). As you suggested,  I'll probably skip these functions.  If so,  I make sure to error handle calls in udp context.

**Comment by agentzh** [comment]
@alonbg Yeah, throwing out a Lua exception in those unsupported Lua APIs is important.

**Comment by agentzh** [comment]
@alonbg Please rebase to the latest master which contains the cjson travis ci fix.

File: t/138-req-udp-socket.t [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+    content_by_lua_block {
+        local sock, err = ngx.req.udp_socket()
+        if sock then
+            ngx.say("got the request socket")
+        else
+            ngx.say("failed to get the request socket: ", err)
+        end
+
+        for i = 1, 3 do
+            local data, err, part = sock:receive(5)
```
Comment (agentzh) [inline t/138-req-udp-socket.t]: The `receive()` method should not take any argument. Datagrams are atomic, unlike streams.

File: t/138-req-udp-socket.t [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+            if data then
+                ngx.say("received: ", data)
```
Comment (agentzh) [inline t/138-req-udp-socket.t]: We should call `sock:send()` instead of `ngx.say()` here.

**Comment by agentzh** [comment]
@alonbg I've just made a dirty and quick patch against your branch that makes the simplest test case pass completely on my side. Just for your reference:

```diff
diff --git a/src/ngx_stream_lua_socket_udp.c b/src/ngx_stream_lua_socket_udp.c
index d066685..366f7a4 100644
--- a/src/ngx_stream_lua_socket_udp.c
+++ b/src/ngx_stream_lua_socket_udp.c
@@ -1058,7 +1058,7 @@ ngx_stream_lua_socket_udp_send(lua_State *L)
 
     dd("sending query %.*s", (int) query.len, query.data);
 
-    n = ngx_send(u->udp_connection.connection, query.data, query.len);
+    n = ngx_udp_send(u->udp_connection.connection, query.data, query.len);
 
     dd("ngx_send returns %d (query len %d)", (int) n, (int) query.len);
 
@@ -1087,6 +1087,7 @@ ngx_stream_lua_socket_udp_receive(lua_State *L)
 {
     ngx_stream_session_t                  *s;
     ngx_stream_lua_socket_udp_upstream_t  *u;
+    ngx_connection_t                      *c;
     ngx_int_t                              rc;
     ngx_stream_lua_ctx_t                  *ctx;
     ngx_stream_lua_co_ctx_t               *coctx;
@@ -1156,7 +1157,19 @@ ngx_stream_lua_socket_udp_receive(lua_State *L)
                    "stream lua udp socket receive buffer size: %uz",
                    u->recv_buf_size);
 
-    rc = ngx_stream_lua_socket_udp_read(s, u);
+    c = u->udp_connection.connection;
+
+    if (u->raw_downstream && !u->connected) {
+        u->received = c->buffer->last - c->buffer->pos;
+        c->buffer->pos =
+            ngx_copy(ngx_stream_lua_socket_udp_buffer, c->buffer->pos, u->received);
+        ngx_stream_lua_socket_udp_handle_success(s, u);
+        u->connected = 1;
+        rc = NGX_OK;
+
+    } else {
+        rc = ngx_stream_lua_socket_udp_read(s, u);
+    }
 
     if (rc == NGX_ERROR) {
         dd("read failed: %d", (int) u->ft_type);
@@ -1268,7 +1281,7 @@ ngx_stream_lua_socket_udp_finalize(ngx_stream_session_t *s,
         u->resolved->ctx = NULL;
     }
 
-    if (u->udp_connection.connection) {
+    if (u->udp_connection.connection && !u->raw_downstream) {
         ngx_log_debug0(NGX_LOG_DEBUG_STREAM, s->connection->log, 0,
                        "stream lua close socket connection");
 
diff --git a/src/ngx_stream_lua_socket_udp.h b/src/ngx_stream_lua_socket_udp.h
index b147b54..e7ca378 100644
--- a/src/ngx_stream_lua_socket_udp.h
+++ b/src/ngx_stream_lua_socket_udp.h
@@ -53,7 +53,8 @@ struct ngx_stream_lua_socket_udp_upstream_s {
 
     ngx_stream_lua_co_ctx_t           *co_ctx;
     unsigned                           raw_downstream:1;
-    unsigned                           waiting; /* :1 */
+    unsigned                           waiting:1;
+    unsigned                           connected:1;
 };
 
 
diff --git a/t/138-req-udp-socket.t b/t/138-req-udp-socket.t
index f143c6e..f2b70fb 100644
--- a/t/138-req-udp-socket.t
+++ b/t/138-req-udp-socket.t
@@ -20,31 +20,33 @@ run_tests();
 __DATA__
 
 === TEST 1: sanity
+--- ONLY
 --- dgram_server_config
     content_by_lua_block {
         local sock, err = ngx.req.udp_socket()
-        if sock then
-            ngx.say("got the request socket")
-        else
-            ngx.say("failed to get the request socket: ", err)
+        if not sock then
+            ngx.log(ngx.ERR, "failed to get the request socket: ", err)
+            return ngx.exit(ngx.ERROR)
         end
 
-        for i = 1, 3 do
-            local data, err, part = sock:receive(5)
-            if data then
-                ngx.say("received: ", data)
-            else
-                ngx.say("failed to receive: ", err, " [", part, "]")
-            end
+        local data, err = sock:receive()
+        if not data then
+            ngx.log(ngx.ERR, "failed to receive: ", err)
+            return ngx.exit(ngx.ERROR)
+        end
+
+        print("data: ", data)
+
+        local ok, err = sock:send("received: " .. data)
+        if not ok then
+            ngx.log(ngx.ERR, "failed to send: ", err)
+            return ngx.exit(ngx.ERROR)
         end
     }
 --- dgram_request chomp
 hello world! my
---- dgram_response
-got the request socket
-received: hello
-received:  worl
-received: d! my
+--- dgram_response chomp
+hello world! my
 --- no_error_log
 [error]
 

```

**Comment by splitice** [comment]
What is the status of this PR, is there any assistance needed?

I came across a need for this capability in one of my projects.

**Comment by agentzh** [comment]
@splitice With my dirty patch above, it is good enough for my DNS name server project already. Though it's not good enough for merging into master :)

**Comment by splitice** [comment]
Cool,

And FYI - my project is to implement a caching DNS Proxy. Seems like I am seeing a pattern out of everyone who wants this patch.

**Comment by agentzh** [comment]
FYI: https://twitter.com/agentzh/status/766062354928955392

For the best result there, I had to enable the trace stitching feature of LuaJIT and also to disable the per-request global environment table creation in ngx_stream_lua_module :) I think we can archive the similar effect by recycling the global environment tables to avoid creating and GC many such tables in the formal implementation of both ngx_stream_lua_module and ngx_http_lua_module ;)

**Comment by agentzh** [comment]
Anyway, I'm gonna deploy my pure Lua DNS nameserver atop this NGINX module to power my own openresty.org domain very soon :)

**Comment by splitice** [comment]
Damn, thats fast. I'm hoping to get at-least that. My ideal target is ~200 - 400K r/s with 4 - 8 workers (assuming high cache hit rate, and some fairly beefy CPUs).

Any chance you plan to opensource your server? It would be very interesting to see how you have architected it vs what I have done so far.

My needs dictate a proxy, rather than a server per-say but there is no reason if open sourced not to extend / re-use if possible,

**Comment by agentzh** [comment]
@splitice Just created a gist containing everything needed to run my Lua DNS server atop NGINX yourself:

https://gist.github.com/agentzh/6c50d37510daef792ed220fa0d970393

It's very hacky though ;)

There is still room for optimizations. See the C-land on-CPU flamegraph while my Lua DNS server atop NGINX is loaded by dnsperf:

http://agentzh.org/misc/flamegraph/lua-dns-server-on-nginx-2016-08-17.svg

Basically with a pure C DNS parser and generator can call them via FFI should be faster. Also, we could recycle the Lua thread object for each coroutine, reducing GC overhead.

Have fun :)

**Comment by alonbg** [comment]
@agentzh absolutely fabulous
@splitice yes, help is needed making it master-merge ready  :)
I also need a udp proxy rather than a terminating server and also to manipulate upstream and downstream.
I also need nginx's [`proxy_bind transparent`](http://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_bind) feature introduced in nginx 1.11.2.
As a result I wouldn't be able to use `content_by_lua_block`

For the upstream manipulation I thought of making [this] (https://github.com/ryeyao/stream-lua-nginx-module)  `access_by_lua_block` feature [work with nginx 1.11.2](https://github.com/ryeyao/stream-lua-nginx-module/issues/3).

And for the downstream I have no direction yet.
It should be something like `filter_datagram_by_lua_block` with an optional parameter for number of datagrams to be captured by the filter. Don't know if that's feasible or not.

An alternative to the above would be to implement the `proxy_bind` feature on the upstream co-sockets. Thus avoiding the us of nginx proxy which in turn would allow the entire downstream/upstream manipulation to occur in `content_by_lua_block` context.
  
Any kind of help you can offer would be great

**Comment by agentzh** [comment]
@alonbg You may find this PR for ngx_http_lua_module interesting: https://github.com/openresty/lua-nginx-module/pull/712 It adds the `bind()` method to cosockets.

**Comment by splitice** [comment]
I don't think ```filter_datagram_by_lua_block``` would be simple to implement, my understanding is that the stream module is structured differently to the the http module.

I'm working under the assumption that (unfortunately) the full manipulation has to occur in the ```content_by_lua_block```.

**Comment by alonbg** [comment]
@agentzh I'll look into the bind PR in ngx_http_lua_module. Thank you.
Though, I do need the `IP_TRANSPARENT` socket option for non-local addresses.
I'll try adding that as well. 

@splitice That's a very important confirmation. Thank you for sharing your thoughts :).  I believe I'll continue working under those same assumptions.

**Comment by alonbg** [comment]
@agentzh adopted your dirty patch :)
passed all udp test cases
rebased

File: src/ngx_stream_lua_socket_udp.c [hunk 12]
```diff
@@ -961,7 +1158,18 @@ ngx_stream_lua_socket_udp_receive(lua_State *L)
...
+    c = u->udp_connection.connection;
+
+    if (u->raw_downstream && !u->connected) {
+        u->received = c->buffer->last - c->buffer->pos;
+        c->buffer->pos = ngx_copy(ngx_stream_lua_socket_udp_buffer,
+                                  c->buffer->pos, u->received);
+         ngx_stream_lua_socket_udp_handle_success(s, u);
+         u->connected = 1;
+         rc = NGX_OK;
+     } else {
```
Comment (agentzh) [inline src/ngx_stream_lua_socket_udp.c]: Style: need a blank line before `} else {`. Please fix other places as well. Thank you.

File: src/ngx_stream_lua_socket_udp.c [hunk 13]
```diff
@@ -1125,6 +1354,7 @@ ngx_stream_lua_socket_udp_read(ngx_stream_session_t *s,
     ssize_t                      n;
 
     c = u->udp_connection.connection;
+
```
Comment (agentzh) [inline src/ngx_stream_lua_socket_udp.c]: Why this extra blank line? Please keep your changes minimal :)

File: src/ngx_stream_lua_socket_udp.c [hunk 14]
```diff
@@ -106,18 +113,206 @@ ngx_stream_lua_inject_socket_udp_api(ngx_log_t *log, lua_State *L)
...
+    /* {{{upstream userdata metatable */
+    lua_pushlightuserdata(L, &ngx_stream_lua_upstream_udata_metatable_key);
     lua_createtable(L, 0 /* narr */, 1 /* nrec */); /* metatable */
     lua_pushcfunction(L, ngx_stream_lua_socket_udp_upstream_destroy);
     lua_setfield(L, -2, "__gc");
     lua_rawset(L, LUA_REGISTRYINDEX);
     /* }}} */
 
+    /* {{{downstream userdata metatable */
+    lua_pushlightuserdata(L, &ngx_stream_lua_downstream_udata_metatable_key);
```
Comment (agentzh) [inline src/ngx_stream_lua_socket_udp.c]: I think we need a separate metatable for downstream udp cosocket objects since some of the upstream udp cosocket object methods like `setpeername` just does not make sense for downstream.

File: src/ngx_stream_lua_output.c [hunk 15]
```diff
@@ -60,6 +60,10 @@ ngx_stream_lua_ngx_echo(lua_State *L, unsigned newline)
         return luaL_error(L, "no session object found");
     }
 
+    if (s->connection->type == SOCK_DGRAM) {
+        return luaL_error(L, "api call is not supported under udp transport");
```
Comment (agentzh) [inline src/ngx_stream_lua_output.c]: Also need such errors for `ngx.flush()` and `ngx.eof()` as well. Maybe we can shorten the message to `"not supported in udp requests"`.

File: src/ngx_stream_lua_socket_udp.c (hunk 14 above)
Comment (alonbg) [inline src/ngx_stream_lua_socket_udp.c]: Unless my understanding is wrong, like the tcp case, udp downstream does have a separate meta table - `ngx_stream_lua_raw_req_socket_metatable_key`

**Comment by alonbg** [comment]
@agentzh this awaits your approval or is there anything else todo on this pull request?
Just to remind that my latest test-nginx pull request is required for CI to pass udp tests.

**Comment by bjne** [comment]
@agentzh Can this be useful for your dns impl: 
https://github.com/vavrusa/ljdns

**Comment by alonbg** [comment]
continues [in a new pull request](https://github.com/openresty/stream-lua-nginx-module/pull/38)
